File: tags/release-0.91/src/gui/parmbox2.cpp
// File:     parmbox2.cpp
// Location: src/gui
// Purpose:  implement ParmBox2 class
// Author:   Stephen Gold sgold@sonic.net
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the 
Free Software Foundation, either version 3 of the License, or (at your 
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifdef _WINDOWS
#include "gui/parmbox2.hpp"
#include "gui/resource.hpp"
#include "gui/win_types.hpp"

// message handler (callback) for this dialog box

static INT_PTR CALLBACK message_handler(
    HWND windowHandle,
    MessageType message,
    WPARAM wParameter, 
    LPARAM lParameter)
{
    lParameter;  // unused parameter
    ASSERT(windowHandle != NULL);
    ParmBox2* const p_box = (ParmBox2*)BaseWindow::Lookup(windowHandle);
    ASSERT(HWND(*p_box) == windowHandle);
    INT_PTR const result = p_box->HandleMessage(message, wParameter);

    return result;
}


// lifecycle

ParmBox2::ParmBox2(GameOpt& rGameOpt):
Dialog("PARMBOX2", &message_handler),
    mrGameOpt(rGameOpt)
{
}


// misc methods

IndexType ParmBox2::GetTextIndex(IdType controlId) {
    String const string = GetTextString(controlId);

    IndexType result = Cell::HEIGHT_MAX;
    if (string != "endless") {
        result = IndexType(string);
    }

    return result;
}

INT_PTR ParmBox2::HandleMessage(MessageType message, WPARAM wParam) {
    INT_PTR result = FALSE;

    switch (message) {
    case WM_INITDIALOG: {
        Dialog::HandleMessage(message, wParam);

        SetGrid();
        SetTopology();

        SetTextIndex(IDC_EDITHEIGHT, mrGameOpt.BoardHeight());
        SetTextIndex(IDC_EDITWIDTH, mrGameOpt.BoardWidth());
        UpdateCellCnt();

        result = TRUE;
        break;
                        }

    case WM_COMMAND: {
        IdType const id = LOWORD(wParam);
        int const notification_code = HIWORD(wParam);
        switch (id) {
        case IDC_EDITHEIGHT: {
            RowType height = RowType(GetTextIndex(id));
            if (::is_odd(height)) { // make even
                height++;
            }
            if (height >= Cell::HEIGHT_MIN
                && height <= Cell::HEIGHT_MAX) {
                    mrGameOpt.SetBoardHeight(height);
            }
            SetTopology();
            UpdateCellCnt();
            break;
                             }
        case IDC_EDITWIDTH: {
            ColumnType width = GetTextIndex(id);
            if (::is_odd(width)) { // make even
                width++;
            }
            if (width >= Cell::WIDTH_MIN
                && width <= Cell::WIDTH_MAX) {
                    mrGameOpt.SetBoardWidth(width);
            }
            SetTopology();
            UpdateCellCnt();
            break;
                            }
        case IDC_RECALC:
            if (notification_code == BN_CLICKED) {
                SetTopology();
                SetTextIndex(IDC_EDITHEIGHT, mrGameOpt.BoardHeight());
                SetTextIndex(IDC_EDITWIDTH, mrGameOpt.BoardWidth());
                UpdateCellCnt();
            }
            break;
        case IDC_TRIANGLE:
        case IDC_4WAY:
        case IDC_HEX:
        case IDC_8WAY:
            if (notification_code == BN_CLICKED) {
                SetGrid(id);
                SetTopology();
                SetTextIndex(IDC_EDITHEIGHT, mrGameOpt.BoardHeight());
                SetTextIndex(IDC_EDITWIDTH, mrGameOpt.BoardWidth());
                UpdateCellCnt();
            }
            break;

        case IDC_ENDLESS:
        case IDC_RECT:
        case IDC_TORUS:
        case IDC_VSTRIP:
        case IDC_HSTRIP:
        case IDC_VCYLINDER:
        case IDC_HCYLINDER:
            if (notification_code == BN_CLICKED) {
                SetTopology(id);
                SetGrid();
                SetTextIndex(IDC_EDITHEIGHT, mrGameOpt.BoardHeight());
                SetTextIndex(IDC_EDITWIDTH, mrGameOpt.BoardWidth());
                UpdateCellCnt();
            }
            break;
        }
        break;
                     }
    }

    if (result == FALSE) {
        result = Dialog::HandleMessage(message, wParam);
    }

    return result;
}

void ParmBox2::SetGrid(void) {
    EnableControl(IDC_TRIANGLE, false); // TODO
    EnableControl(IDC_4WAY, true);
    EnableControl(IDC_HEX, true);
    EnableControl(IDC_8WAY, true);

    GridType const grid = GridType(mrGameOpt);
    SetButton(IDC_TRIANGLE, grid == GRID_TRIANGLE);
    SetButton(IDC_4WAY, grid == GRID_4WAY);
    SetButton(IDC_HEX, grid == GRID_HEX);
    SetButton(IDC_8WAY, grid == GRID_8WAY);    
}

void ParmBox2::SetGrid(IdType buttonId) {
    switch (buttonId) {
    case IDC_TRIANGLE:
        mrGameOpt.SetGrid(GRID_TRIANGLE);
        break;
    case IDC_4WAY:
        mrGameOpt.SetGrid(GRID_4WAY);
        break;
    case IDC_HEX:
        mrGameOpt.SetGrid(GRID_HEX);
        break;
    case IDC_8WAY:
        mrGameOpt.SetGrid(GRID_8WAY);
        break;
    default:
        FAIL();
    }
}

void ParmBox2::SetTextIndex(IdType controlId, ValueType value) {
    if (value == Cell::HEIGHT_MAX) {
        SetTextString(controlId, "endless");
    } else {
        SetTextValue(controlId, value);
    }
}

void ParmBox2::SetTopology(void) {
    EnableControl(IDC_EDITHEIGHT, true);
    EnableControl(IDC_EDITWIDTH, true);
    EnableControl(IDC_ENDLESS, true);
    EnableControl(IDC_RECT, true);
    EnableControl(IDC_HSTRIP, true);
    EnableControl(IDC_VSTRIP, true);
    EnableControl(IDC_TORUS, true);
    EnableControl(IDC_VCYLINDER, true);
    EnableControl(IDC_HCYLINDER, true);

    bool const wrap = mrGameOpt.DoesBoardWrap();
    bool const finite_height = mrGameOpt.HasFiniteHeight();
    bool const finite_width = mrGameOpt.HasFiniteWidth();
    bool const inf_height = !finite_height;
    bool const inf_width = !finite_width;

    SetButton(IDC_ENDLESS, inf_height && inf_width);
    SetButton(IDC_RECT, !wrap && finite_height && finite_width);
    SetButton(IDC_TORUS, wrap && finite_height && finite_width);
    SetButton(IDC_VSTRIP, !wrap && inf_height && finite_width);
    SetButton(IDC_HSTRIP, !wrap && finite_height && inf_width);
    SetButton(IDC_VCYLINDER, wrap && inf_height && finite_width);
    SetButton(IDC_HCYLINDER, wrap && finite_height && inf_width);
}

void ParmBox2::SetTopology(IdType buttonId) {
    // update board height and width
    RowType height = mrGameOpt.BoardHeight();
    ColumnType width = mrGameOpt.BoardWidth();

    switch (buttonId) {
    case IDC_ENDLESS:
        height = Cell::HEIGHT_MAX;
        width = Cell::WIDTH_MAX;
        break;

    case IDC_RECT:
    case IDC_TORUS:
        if (height == Cell::HEIGHT_MAX) {
            height = HEIGHT_DEFAULT;
        }
        if (width == Cell::WIDTH_MAX) {
            width = WIDTH_DEFAULT;
        }
        break;

    case IDC_VSTRIP:
    case IDC_VCYLINDER:
        height = Cell::HEIGHT_MAX;
        if (width == Cell::WIDTH_MAX) {
            width = WIDTH_DEFAULT;
        }
        break;

    case IDC_HSTRIP:
    case IDC_HCYLINDER:
        if (height == Cell::HEIGHT_MAX) {
            height = HEIGHT_DEFAULT;
        }
        width = Cell::WIDTH_MAX;
        break;

    default:
        FAIL();
    }

    mrGameOpt.SetBoardHeight(height);
    mrGameOpt.SetBoardWidth(width);

    // update wrap flag
    bool does_wrap = mrGameOpt.DoesBoardWrap();
    switch (buttonId) {
    case IDC_ENDLESS:
        // mWrapFlag doesn't matter
        break;

    case IDC_HSTRIP:
    case IDC_RECT:
    case IDC_VSTRIP:
        does_wrap = false;
        break;

    case IDC_HCYLINDER:
    case IDC_TORUS:
    case IDC_VCYLINDER:
        does_wrap = true;
        break;

    default:
        FAIL();
    }
    mrGameOpt.SetDoesBoardWrap(does_wrap);
}

void ParmBox2::UpdateCellCnt(void) {
    IndexType cell_cnt = Cell::HEIGHT_MAX;

    if (mrGameOpt.HasFiniteHeight() && mrGameOpt.HasFiniteWidth()) {
        RowType const height = mrGameOpt.BoardHeight();
        ColumnType const width = mrGameOpt.BoardWidth();
        if (GridType(mrGameOpt) == GRID_HEX) { // hex grids have half as many cells
            cell_cnt = (height/2) * width;
        } else {
            cell_cnt = height * width;
        }
    }
    SetTextIndex(IDC_CELLCNT, cell_cnt);
}
#endif // defined(_WINDOWS)

File: Java/GoldTile/src/goldtile/ParmBox2.java
// File:     ParmBox2.java
// Location: Java/GoldTile/src/goldtile
// Purpose:  ParmBox2 class for the Gold Tile Game
/**
 * @author Stephen Gold
 */
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the 
Free Software Foundation, either version 3 of the License, or (at your 
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

package goldtile;

import java.awt.event.ActionListener;
import javax.swing.*;
import javax.swing.border.Border;
import javax.swing.border.EtchedBorder;

public class ParmBox2
    extends WizardCard
    implements ActionListener // radio buttons and Dim fields
{
    // constants
    final private static int GRID_BUTTON_COUNT = 4;
    final private static int TOPOLOGY_BUTTON_COUNT = 7;
    
    // embedded classes, sorted by name
    
    private class DimField extends JTextField {
        DimField(ActionListener listener) {
            super(Dim.endlessString(), 10); // 10 characters wide
            setHorizontalAlignment(JTextField.RIGHT);
            
            // change font to match JLabel
            final java.awt.Font labelFont = new JLabel().getFont();
            setFont(labelFont);
            
            addActionListener(listener);
            setInputVerifier(new DimVerifier());
            
            setMaximumSize(new java.awt.Dimension(70, 30));
        }
        
        public Dim getValue() {
            final String text = getText();
            return new Dim(text);
        }
        
        public void setValue(Dim value) {
            final String text = value.toString();
            setText(text);
        }
    }
    
    public class GridIcon implements Icon {
        // methods, sorted by name
        @Override
        public int getIconHeight() { return 90; }
        @Override
        public int getIconWidth() { return 90; }
        @Override
        public void paintIcon(java.awt.Component component, 
                java.awt.Graphics graphics, int x, int y)
        {
            // TODO
        }
    }
    
    private static class InvalidGridException extends Exception {}
    private static class InvalidTopologyException extends Exception {}
    
    // per-instance fields, sorted by type
    final private DimField heightField;
    final private DimField widthField;
    private GameOpt gameOpt = null;
    private HandOpt[] handOpts = null;
    final private JLabel cellsLabel = new JLabel("TODO cells");
    final private JRadioButton gridButtons[] = 
            new JRadioButton[GRID_BUTTON_COUNT];
    final private JRadioButton topologyButtons[] = 
            new JRadioButton[TOPOLOGY_BUTTON_COUNT];

    // constructors
    
    @SuppressWarnings("LeakingThisInConstructor")
    public ParmBox2(Wizard wizard) {
        super(wizard, ParmBox2.class.getName());
        
        // board types
        
        final Box topologyBoxLeft = new Box(BoxLayout.PAGE_AXIS);
        topologyBoxLeft.setAlignmentY(Box.TOP_ALIGNMENT);
        final Box topologyBoxRight = new Box(BoxLayout.PAGE_AXIS);
        topologyBoxRight.setAlignmentY(Box.TOP_ALIGNMENT);
        final ButtonGroup topologyGroup = new ButtonGroup(); 
        
        for (Topology topology : Topology.values()) {
            try {
                final int iButton = getButtonIndex(topology);
                final JRadioButton button = new JRadioButton();
                button.setText(getButtonText(topology));
                button.setActionCommand(topology.name());
                button.addActionListener(this);
                button.setAlignmentX(JRadioButton.LEFT_ALIGNMENT);
            
                if (iButton < 3) {
                    topologyBoxLeft.add(button);
                    if (iButton < 2) {
                        topologyBoxLeft.add(Box.createVerticalGlue());
                    }
                } else {
                    topologyBoxRight.add(button);
                }
                topologyGroup.add(button);
                topologyButtons[iButton] = button;
            } catch (InvalidTopologyException exception) {
            }
        }
        
        final Box topologyBox = new Box(BoxLayout.LINE_AXIS);
        final Border topologyBorder = BorderFactory.createTitledBorder( 
                BorderFactory.createEtchedBorder(EtchedBorder.LOWERED),
                "Which sort of board would you prefer?" );
        topologyBox.setBorder(topologyBorder);
        topologyBox.add(topologyBoxLeft);
        topologyBox.add(Box.createHorizontalGlue());
        topologyBox.add(topologyBoxRight);
        
        // grid types
        
        final Box gridBox = new Box(BoxLayout.PAGE_AXIS);
        final ButtonGroup gridGroup = new ButtonGroup();
        
        for (Grid grid : Grid.values()) {
            try {
                final int iButton = getButtonIndex(grid);
                final JRadioButton button = new JRadioButton();
                button.setText(getButtonText(grid));
                button.setActionCommand(grid.name());
                button.addActionListener(this);
                button.setAlignmentX(JRadioButton.LEFT_ALIGNMENT);
            
                gridBox.add(button);
                gridGroup.add(button);
                gridButtons[iButton] = button;
            } catch (InvalidGridException exception) {
            }
        }
        
        final Box lowerLeft = new Box(BoxLayout.LINE_AXIS);
        final Border gridBorder = BorderFactory.createTitledBorder( 
                BorderFactory.createEtchedBorder(EtchedBorder.LOWERED),
                "Which sort of grid would you prefer?" );
        lowerLeft.setBorder(gridBorder);
        lowerLeft.add(gridBox);
        final JLabel gridImage = new JLabel(new GridIcon());
        lowerLeft.add(gridImage);        
        
        // board width/height textfields with custom verifiers

        final Box widthBox = new Box(BoxLayout.LINE_AXIS);
        widthBox.setAlignmentX(JRadioButton.CENTER_ALIGNMENT);
        widthBox.add(new JLabel("Board width: "));
        widthField = new DimField(this);
        widthBox.add(widthField);
        widthBox.add(new JLabel(" columns"));
        
        final Box heightBox = new Box(BoxLayout.LINE_AXIS);
        heightBox.setAlignmentX(JRadioButton.CENTER_ALIGNMENT);
        heightBox.add(new JLabel("Board height: "));
        heightField = new DimField(this);
        heightBox.add(heightField);
        heightBox.add(new JLabel(" rows"));
        
        cellsLabel.setAlignmentX(JRadioButton.CENTER_ALIGNMENT);

        // put it all together
        
        final Box bottomRight = new Box(BoxLayout.PAGE_AXIS);
        bottomRight.add(Box.createVerticalGlue());
        bottomRight.add(widthBox);
        bottomRight.add(heightBox);
        bottomRight.add(Box.createVerticalGlue());
        bottomRight.add(cellsLabel);
        bottomRight.add(Box.createVerticalGlue());
        
        final Box bottom = new Box(BoxLayout.LINE_AXIS);
        bottom.add(lowerLeft);
        bottom.add(bottomRight);
        
        final Box mainBox = new Box(BoxLayout.PAGE_AXIS);
        mainBox.add(topologyBox);
        mainBox.add(bottom);
        
        add(mainBox);
    }
    
    // methods, sorted by name

    @Override
    public void actionPerformed(java.awt.event.ActionEvent event) {
        assert event != null;
        
        updateModel();

        final String command = event.getActionCommand();        
        try {
            final Grid grid = Grid.valueOf(command);
            
            // from a grid button        
            gameOpt.setGrid(grid);
            
        } catch (IllegalArgumentException exception) {
        }
        
        try {
            final Topology topology = Topology.valueOf(command);

            // from a topology button        
            gameOpt.setBoardWrap(topology.doesWrap());
            
            if (topology.hasFiniteHeight() != gameOpt.hasFiniteHeight()) {
                if (topology.hasFiniteHeight()) {
                    gameOpt.setBoardHeight(GameOpt.BOARD_HEIGHT_DEFAULT);
                } else {
                    gameOpt.setBoardHeight(Dim.endless());
                }
            }

            if (topology.hasFiniteWidth() != gameOpt.hasFiniteWidth()) {
                if (topology.hasFiniteWidth()) {
                    gameOpt.setBoardWidth(GameOpt.BOARD_WIDTH_DEFAULT);
                } else {
                    gameOpt.setBoardWidth(Dim.endless());
                }
            }

        } catch (IllegalArgumentException exception) {
        }
 
        updateView();
    }
    
    @Override
    public void backAction() {
        updateModel();
        showCard(ParmBox1.class.getName(), gameOpt, handOpts);            
    }

    private int getButtonIndex(Grid grid) 
            throws InvalidGridException
    {
        switch (grid) {
            case GRID_4WAY:
                return 0;
            case GRID_8WAY:
                return 1;
            case GRID_HEX:
                return 2;
            case GRID_TRIANGLE:
                return 3;
            default:
                throw new InvalidGridException();
        }
    }

    private int getButtonIndex(Topology topology) 
            throws InvalidTopologyException
    {
        switch (topology) {
            case FLAT:
                return 0;
            case RECTANGLE:
                return 1;
            case TORUS:
                return 2;
            case VSTRIP:
                return 3;
            case HSTRIP:
                return 4;
            case VCYLINDER:
                return 5;
            case HCYLINDER:
                return 6;
            default:
                throw new InvalidTopologyException();
        }
    }

    private String getButtonText(Grid grid) {
        switch (grid) {
            case GRID_4WAY:
                return "Square grid (4-way)";
            case GRID_8WAY:
                return "Square grid (8-way)";
            case GRID_HEX:
                return "Hexagonal grid";
            case GRID_TRIANGLE:
                return "Triangular grid";
            default:
                throw new AssertionError(grid);
        }
    }
    
    private String getButtonText(Topology topology) {
        switch (topology) {
            case FLAT:
                return "Flat and endless (no wrap)";
            case RECTANGLE:
                return "Rectangle/square (4 edges)";
            case TORUS:
                return "Torus (wraps in all directions)";
            case VSTRIP:
                return "Vertical strip (left/right edges)";
            case HSTRIP:
                return "Horizontal strip (top/bottom edges)";
            case VCYLINDER:
                return "Vertical cylinder (wraps left/right)";
            case HCYLINDER:
                return "Horizontal cylinder (wraps top/bottom)";
            default:
                throw new AssertionError(topology);
        }
    }
    
    @Override 
    public String getTitle() {
        return "Board Parameters";
    }
    
    @Override
    public void nextAction() {
        updateModel();
        
        showCard(ParmBox3.class.getName(), gameOpt, handOpts);
    }
    
    @Override
    public void setModels(Object[] models) {
        assert models != null;
        assert models.length == 2 : models.length;
        assert models[0] != null;
        assert models[1] != null;
        
        gameOpt = (GameOpt)models[0];
        gameOpt.validate();
        handOpts = (HandOpt[])models[1];
        
        updateView();
    }
    
    private void updateModel() {
        final Dim height = heightField.getValue();
        gameOpt.setBoardHeight(height);

        final Dim width = widthField.getValue();
        gameOpt.setBoardWidth(width);

        gameOpt.validate();
    }
    
    public void updateView() {
        // the wizard's back/next buttons
        wizard.getBackButton().setText("\u21d0 Back");
        wizard.getBackButton().setEnabled(true);
        wizard.getBackButton().setVisible(true);
        wizard.getNextButton().setText("Next \u21d2");
        wizard.getNextButton().setEnabled(true);
        
        // button values
        
        final Grid grid = gameOpt.getGrid();
        try {
            final int iButton = getButtonIndex(grid);
            gridButtons[iButton].setSelected(true);    
        } catch (InvalidGridException exception) {
            throw new AssertionError();
        }
        
        final Topology topology = gameOpt.getTopology();
        try {
            final int iButton = getButtonIndex(topology);
            topologyButtons[iButton].setSelected(true);    
        } catch (InvalidTopologyException exception) {
            throw new AssertionError();
        }
        
        // dimensions
        final Dim height = gameOpt.getBoardHeight();
        heightField.setValue(height);
        final Dim width = gameOpt.getBoardWidth();
        widthField.setValue(width);
        
        // cell count
        final long cellCount = gameOpt.getCellCount();
        String text = Long.toString(cellCount);
        if (cellCount == Long.MAX_VALUE) {
            text = Dim.endlessString();
        }
        cellsLabel.setText(String.format("\u21d2 The board has %s cells.", 
                    text));            
    }    
}
